hivelocity/instances_v2.go

```diff
@@ -0,0 +1,162 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hivelocity
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/hivelocity/hivelocity-cloud-controller-manager/client"
+	"github.com/hivelocity/hivelocity-cloud-controller-manager/pkg/hvutils"
+	corev1 "k8s.io/api/core/v1"
+	cloudprovider "k8s.io/cloud-provider"
+)
+
+// HVInstancesV2 implements cloudprovider.InstanceV2.
+type HVInstancesV2 struct {
+	client client.Interface
+}
+
+var _ cloudprovider.InstancesV2 = &HVInstancesV2{}
+
+// newHVInstanceV2 creates a new HVInstancesV2 struct.
+func newHVInstanceV2(c client.Interface) *HVInstancesV2 {
+	return &HVInstancesV2{client: c}
+}
+
+// getHivelocityDeviceIDFromNode returns the deviceID from a Node.
+// Example: If Node.Spec.ProviderID is "hivelocity://123", then 123
+// will be returned.
+func getHivelocityDeviceIDFromNode(node *corev1.Node) (int32, error) {
+	providerPrefix := providerName + "://"
+	if !strings.HasPrefix(node.Spec.ProviderID, providerPrefix) {
+		return 0, fmt.Errorf("node: %s, ProviderID: %q: %w", node.GetName(), node.Spec.ProviderID,
+			errMissingProviderPrefix)
+	}
+	deviceID, err := strconv.ParseInt(strings.TrimPrefix(node.Spec.ProviderID, providerPrefix), 10, 32) //nolint:revive // https://www.reddit.com/r/golang/comments/10mi0el/strconvparseintmyid_10_32_avoid_magic_numbers/
+	if err != nil {
+		return 0, fmt.Errorf("node: %s, ProviderID %q: %w",
+			node.GetName(), node.Spec.ProviderID, errFailedToConvertProviderID)
+	}
+	return int32(deviceID), nil
+}
+
+var (
+	errMissingProviderPrefix     = fmt.Errorf("missing prefix %q in node.Spec.ProviderID", providerName)
+	errFailedToConvertProviderID = fmt.Errorf("failed to convert node.Spec.ProviderID")
+)
+
+// InstanceExists returns true if the instance for the given node exists according to the cloud provider.
+// Use the node.name or node.spec.providerID field to find the node in the cloud provider.
+// Implements cloudprovider.InstancesV2.InstanceExists.
+func (i2 *HVInstancesV2) InstanceExists(ctx context.Context, node *corev1.Node) (bool, error) {
+	if node == nil {
+		return false, ErrNodeIsNil
+	}
+	deviceID, err := getHivelocityDeviceIDFromNode(node)
+	if err != nil {
+		return false, fmt.Errorf("InstanceExists(): getHivelocityDeviceIDFromNode() failed. %w", err)
+	}
+	_, err = i2.client.GetBareMetalDevice(ctx, deviceID)
+	if errors.Is(err, client.ErrNoSuchDevice) {
+		return false, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("InstanceExists(): GetBareMetalDevice() failed. deviceID %d: %w", deviceID, err)
+	}
+	return true, nil
+}
+
+// ErrUnknownPowerStatus .
+var ErrUnknownPowerStatus = errors.New("unknown PowerStatus")
```

Comment:
why do you export all these errors? If you didn't, the linter also wouldn't force you to describe them

Reply:
fixed. these are now unexported.